server/controllers/finance/reports/cashflow/index.js

```diff
@@ -753,3 +169,241 @@ function reportByService(req, res, next) {
+  const dateTo = new Date(req.query.dateTo);
+  const options = _.clone(req.query);
+  const data = {};
+
+  // convert cashboxesIds parameters in array format ['', '', ...]
+  // this parameter can be sent as a string or an array we force the conversion into an array
+  const cashboxesIds = _.values(req.query.cashboxesIds);
+
+  _.extend(options, { orientation : 'landscape' });
+
+  // catch missing required parameters
+  if (!dateFrom || !dateTo || !cashboxesIds.length) {
+    throw new BadRequest('ERRORS.BAD_REQUEST');
+  }
+
+  try {
+    serviceReport = new ReportManager(TEMPLATE, req.session, options);
+  } catch (e) {
+    next(e);
+    return;
+  }
+
+  data.dateFrom = dateFrom;
+  data.dateTo = dateTo;
+
+  getCashboxesDetails(cashboxesIds)
+    .then(rows => {
+      data.cashboxes = rows;
+
+      data.cashAccountIds = data.cashboxes.map(cashbox => cashbox.account_id);
+
+      data.cashLabels = _.chain(data.cashboxes)
+        .map(cashbox => `${cashbox.label}`).uniq().join(' | ')
+        .value();
+
+      data.cashLabelSymbol = _.chain(data.cashboxes)
+        .map(cashbox => cashbox.symbol).uniq().join(' + ');
+
+      data.cashLabelDetails = data.cashboxes.map(cashbox => `${cashbox.account_number} - ${cashbox.account_label}`);
+
+      // get oposite lines to cashbox accounts in transactions as details
+      return getDetailsIdentifiers(data.cashAccountIds, dateFrom, dateTo);
+    })
+    .then(rows => {
+      data.detailsIdentifiers = rows.map(row => row.uuid);
+
+      // build periods columns from calculated period
+      return Fiscal.getPeriodsFromDateRange(data.dateFrom, data.dateTo);
+    })
+    .then(periods => {
+      data.periodDates = periods.map(p => p.start_date);
+
+      data.periods = periods.map(p => p.id);
+
+      data.colspan = data.periods.length + 1;
+
+      // skip period matrix if not transactions were identified
+      if (data.detailsIdentifiers.length === 0) { return []; }
+
+      // build periods string for query
+      const periodString = data.periods.length ? data.periods.map(periodId => {
+        return `SUM(CASE WHEN source.period_id = ${periodId} THEN source.balance ELSE 0 END) AS "${periodId}"`;
+      }).join(',') : '"NO_PERIOD" AS period';
+
+      const query = `
+        SELECT 
+          source.transaction_text, source.account_label, ${periodString},
+          source.transaction_type, source.transaction_id, source.account_id
+        FROM (
+          SELECT
+            a.number AS account_number, a.label AS account_label,
+            SUM(gl.debit_equiv - gl.credit_equiv) AS balance,
+            tt.id AS transaction_id, tt.type AS transaction_type,
+            MAX(tt.text) AS transaction_text,
+            gl.period_id, gl.account_id
+          FROM general_ledger gl
+          JOIN account a ON gl.account_id = a.id
+          JOIN transaction_type tt ON gl.origin_id = tt.id
+          WHERE gl.uuid IN ? 
+          GROUP BY a.id, tt.id
```

Comment:
Do you need this GROUP BY?  It looks like you only need to do one GROUP BY - the outer one.  What happens if you remove the `SUM(gl.debit_equiv - gl.credit_equiv) as balance` and just use `(gl.debit_equiv - gl.credit_equiv) as balance`?   Would that give you the same result?

Reply:
Yes, it is required because the query have to split accounts by transaction type (transaction_type.text : purchase, sale, cash_payment, ...).

We have to notice that the `transaction_type` which is outside correspond to transaction_type.type which can be income, expanse or transfer.

If we remove this `GROUP BY a.id, tt.id` we will get different results.